content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,129 @@
+---
+date: "2020-10-15"
+title: "Announcing Tokio 0.3 and the path to 1.0"
+description: "October 15, 2020"
+---
+
+The Tokio team is excited to announce the release of Tokio 0.3. This release functions as a Tokio 1.0 beta. API rough edges have been fixed. This release is an opportunity to validate the changes before stabilizing them as part of the 1.0 release. As most of these issues were small, we expect that upgrading from 0.2 to 0.3 will be easy.
+
+## Plan for 1.0
+
+The Tokio team plans to release Tokio 1.0 by the end of December 2020. This date is fast approaching. We ask that *you*, the Tokio community, try out Tokio 0.3 and give us feedback through [GitHub Issues](https://github.com/tokio-rs/tokio/issues) or our [Discord channel](https://discord.gg/tokio). Once we release Tokio 1.0, we will commit to the following stability guarantees:
+
+1. A minimum of 5 years of maintenance.
+2. A minimum of 3 years before a hypothetical 2.0 release.
+
+## What’s new?
+
+The main changes in Tokio 0.3 are:
+
+1. Changes to IO traits.
+2. New runtime builder.
+3. The I/O driver is overhauled
+4. The API is future proofed.
+
+You can find the full list on the [changelog](TODO LINK).
+
+## Changes to IO traits
+
+Following the lead of @sfackler’s [RFC](https://github.com/rust-lang/rfcs/pull/2930), we changed the `AsyncRead` and `AsyncWrite` traits to support reading into uninitialized memory. This change *only* affects users who implement the `AsyncRead` or `AsyncWrite` traits or manually calling the `poll_*` methods. If the `AsyncReadExt` or `AsyncWriteExt` traits are used, no change is required.
+
+Below is a simplified version of the new `AsyncRead`:
+
+```rust
+pub trait AsyncRead {
+    fn poll_read(
+        self: Pin<&mut Self>, 
+        cx: &mut Context<'_>, 
+        buf: &mut ReadBuf<'_>
+    ) -> Poll<Result<()>>;
+}
+
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    filled: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    // functions here, see RFC
+}
+```
+The `&mut [u8]` argument to the Tokio 0.2’s `poll_read` method has some unintended sharp edges: an implementer, not the end consumer, of the `AsyncRead` trait can *read* the data stored in the `&mut [u8]` slice. If a programmer creates a mutable slice of bytes (a `&mut [u8]`) that references uninitialized memory (e.g., the excess capacity in a vector), then *reading* the `&mut [u8]` data would cause undefined behavior. The new design mitigates this issue by providing a `ReadBuf` struct that tracks the uninitialized memory, removing the need to initialize memory.
```

Comment:
```suggestion
The `&mut [u8]` argument to Tokio 0.2’s `poll_read` method has some unintended sharp edges: an implementer, not the end consumer, of the `AsyncRead` trait can *read* the data stored in the `&mut [u8]` slice. If a programmer creates a mutable slice of bytes (a `&mut [u8]`) that references uninitialized memory (e.g., the excess capacity in a vector), then *reading* the `&mut [u8]` data would cause undefined behavior. The new design mitigates this issue by providing a `ReadBuf` struct that tracks the uninitialized memory, removing the need to initialize memory.
```